utils/labels.hh

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <seastar/core/metrics.hh>
+
+extern seastar::metrics::label level_label;
+/*!
+ * \brief we use label to allow granularity when reporting metrics
+ */
+
+/*!
+ * \brief Basic level are metrics that are been used by Scylla Monitoring
+ *
+ */
+extern seastar::metrics::label_instance basic_level;
+/*!
+ * \brief Advanced level are metrics that are not part of the dashboard
+ *  but could be helpful at specific times for debugging.
```

Comment:
> The code of this series looks good. I very much agree that enabling and disabling per-table metrics, in particular, should be a query-time decision by Prometheus (who queries Scylla with the Prometheus protocol), and not a Scylla configuration parameter.
> 
> But I'm having a hard time understanding the rationale for the specific choices made in this series. I didn't see an issue or other discussion that led to this design, so I don't know what led to these choices. So for now I'm not approving or rejection this series - just making a comment:
> 
> In particular, I have the following two concerns:
> 
> My first concern is that after this patch, the knowledge of which metrics are more "important" than others will be split into two main places: 1. All over the Scylla source code with these "basic_level" tags, 2. In the Grafana dashboards where we decide which metrics we put in dashboards, and which we don't. It's very easy for the two sources of knowledge to diverge. I can easily imagine myself adding a metric for Alternator and asking you to plot it but then you can't, because I forgot to mark it "basic_level". It seems to me that this knowledge should only exist in one place. For example, our Grafana install could build a list of 50 metrics it wants (because they exist in their metrics), and configure Prometheus to request only these 50 specific metrics instead of all the metrics. If you think it's inefficient to pass the names of 50 metrics on every request, it could be done in two steps (like prepared statements in CQL), or, we can go with an explicit configuration file that needs to be copied to the Scylla server to decide which metrics are important.
> 
> My second concern is what exactly will happen to "non-basic" metrics (I'm not sure, by the way, why the extra levels are important, I didn't see you use them in this series?). If they don't get sent to Prometheus, the users can't view them - not even in "advanced debugging situations". Do you mean in this situation you'll reconfigure Prometheus (not Scylla) to ask for additional labels so the user can see the metrics starting from now? How would this re-configuration be done in practice - do we have a way in Grafana to reconfigure Prometheus, or at least for now the user would ssh into the monitoring node and change some configuration file?

First, you are raising very good questions. They would have been excellent questions had I had good answers.
 
Here's how I see it, there are currently three types of metrics (there's a fourth one that I'll refer to at the end).
The first kind is all the metrics we currently use in the dashboard. The assumption is that those metrics are important enough to be collected and stored all the time. Continued history is essential.

The second kind is all the metrics we are currently reporting but do not display in the dashboard.
Those additional metrics will likely be helpful when someone debugs a system, so we can start collecting those additional metrics at those specific times.
We're in the hand-waving part. One option is to run another Prometheus server when debugging. In the cloud, it could run on a strong machine. Once done, you stop the additional server.
 
The third kind of metrics have labels with high cardinality (like per table metrics, but more may be added in the future).
 I assume that we'll rarely want to collect everything all at once. So if you're going to check a keyspace, a table, or theoretically even an SSTable, you will look at that specific metrics for a duration of time, but it will be specific.
Hand-waving harder, I was thinking about adding an option for an ad-hoc Prometheus-like querier. It would query Scylla for specific metrics on demand. It will have short-term memory and will not use the disk.

The fourth kind, BTW are metrics that we may add in the future that will be useful for specific situations but have additional overhead, the sort of metrics we normally add a specific configuration for.

Last, the part about how we move metrics between the first and second groups?

This is something that I would add to the configuration so we'll be able to make this decision on the fly and later update the code.

A good question is, why not put all the 150 (yup, that's the number) metrics in the configuration?
Mainly because there are so many, it may make more sense to add them to the code.
As we advance, we'll need to think, are we going to collect this new metric, or are we going to just have it there for a rainy day.